Non-deterministic initialisation of structs, pointers and globals in function call harness

NlightNFotis opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Refactor function call harness generator

Important refactoring to make implementation
of later features easier.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add option to non-det globals to goto-harness

Add ability to non-deterministically initialise
globals in the function call harness generator.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add recursive initialisation of structs

Recursively initialise non-deterministically
struct components and pointers.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

hannes-steffenhagen-diffblue commented:
Basically, this PR is intended to re-implement (and ultimately replace) the changes made in https://github.com/diffblue/cbmc/pull/3251, except in an external tool rather than within CBMC

**EDIT**: The diff of the refactoring PR is kind of hopeless, mostly what's going on there is that we break up the generate function into multiple utility functions

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -120,10 +128,29 @@ void function_call_harness_generatort::implt::generate(
+    {
+      const auto &symbol = symbol_table_entry.second;
+      if(
+        symbol.is_static_lifetime && symbol.is_lvalue &&
```
We _think_ this means "global", but we weren't entirely sure. This is basically reverse engineered from a `--show-symbol-table`.

tautschnig replied:
Looks good to me.

xbauch commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -0,0 +1,66 @@
+  /// exist already.
+  symbol_exprt get_malloc_function();
+
+  void get_struct_tag_initializer(
```
Why are these called `get_*` if they don't return?

hannes-steffenhagen-diffblue replied:
Good point, we'll change it

xbauch commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -0,0 +1,126 @@
+  exprt choice;
+  if(
+    depth > initialization_config.min_null_tree_depth &&
+    depth < initialization_config.max_nondet_tree_depth)
```
What about wrapping these conditions in something a bit more descriptive?, e.g.: `should_null_initialize(depth)`, where
```bool should_null_initialize(size_t depth) { return depth >= initialization_config.min_null_tree_depth; }```

hannes-steffenhagen-diffblue replied:
We'll consider it, I think a function might be a bit overkill though. A local variable might be an improvement here though!

xbauch commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -42,6 +72,22 @@ void function_call_harness_generatort::handle_option(
+  {
+    auto const value = require_exactly_one_value(option, values);
+    p_impl->recursive_initialization_config.min_null_tree_depth =
+      std::stoul(value);
```
Don't we have some facility to extract numbers from string safely via `std::optional`?

hannes-steffenhagen-diffblue replied:
We probably do, we'll see how we can improve these

hannes-steffenhagen-diffblue replied:
So, I still think this would be a good idea but we can’t find it. There are the `safe_` conversion functions, but those just assert that the conversion succeeds rather than returning an `optionalt`. Writing new converters on the other hand seems a bit out of scope for this

xbauch replied:
Well `stoul` is safe: throwing exceptions on failure. May want to catch those.

tautschnig replied:
I'd use `safe_string2size_t` here, unless you actually want to throw an exception.

hannes-steffenhagen-diffblue replied:
@tautschnig Considering this is user input I think throwing an exception here would be appropriate. I don't think asserting that user input is always correct is appropriate.

hannes-steffenhagen-diffblue replied:
@xbauch @tautschnig I added https://github.com/diffblue/cbmc/pull/4192, which adds a more convenient way to convert strings to integers and handling failures locally.

tautschnig replied:
That's great, indeed we should not be failing an assertion when reporting invalid user input.

xbauch commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -53,82 +99,147 @@ void function_call_harness_generatort::generate(
+  recursive_initialization->initialize(lhs, 0, block);
+}
+
+void function_call_harness_generatort::validate_options()
```
Is this one called anywhere?

hannes-steffenhagen-diffblue replied:
This is called by the harness generator factory

xbauch commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -53,82 +99,147 @@ void function_call_harness_generatort::generate(
+void function_call_harness_generatort::implt::
+  add_harness_function_to_goto_model(code_blockt function_body)
+{
+  const auto &function_to_call = symbol_table->lookup_ref(function);
```
Why not use `lookup_function_to_call`?

xbauch reviewed (approved):
Looks good.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cac802686f32fbc60d4c268207515e24b07b8669).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100932721
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 365a914eed9b534f7f53236166a9e98a04896fcc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100935342

tautschnig reviewed (approved):
Can we actually get rid of `goto-instrument/nondet_static.cpp` with this code? I think that code might be a bit cleaner when it comes to choosing what (not) to initialise, but isn't very useful when it comes to objects with pointers and such.

chrisr-diffblue reviewed (approved):
LGTM, though you may want to check Michael's comment about using `safe_string2size_t`

danpoe commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -0,0 +1,118 @@
+    allocate_objects.allocate_automatic_local_object(type.subtype(), "pointee");
+  allocate_objects.declare_created_symbols(body);
+  body.add(code_assignt{lhs, null_pointer_exprt{type}});
+  if(depth < initialization_config.max_nondet_tree_depth)
```
This changes the behavior compared to the existing `c_nondet_symbol_factory()`. Previously we would only stop initializing when a recursion on a struct was reached (i.e., on a pointer chain the same struct appeared a second time) and the depth was greater or equal to `max_nondet_tree_depth`. This was done via recording the struct tags in the object `recursion_set`. All pointer dereferences would increase the depth but only when we see a recursion we would stop initializing. Would it be possible to replicate this behavior here? The intention was to be consistent with how the initialization is done in the Java object factory (to facilitate future refactorings and code sharing between the two).

hannes-steffenhagen-diffblue replied:
Ah right, I figured there was something we forgot! Good point, we will implement this.

danpoe commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -53,82 +99,147 @@ void function_call_harness_generatort::generate(
 
-  code_function_callt::operandst arguments{};
-  arguments.reserve(parameters.size());
+void function_call_harness_generatort::implt::generate_nondet_globals(
```
There's also `--nondet-static` (using `nondet_static()`) in cbmc and goto-instrument. I'm not sure how exactly it compares to this functionality but maybe the two could be merged at some point.

NlightNFotis committed:
Only apply tree limit to recursive struct

Now the depth limit only applies to structs if we have seen this struct
before (i.e. the initialisation chain is recursive)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 550e6abfc6fe9579f0c92ef650969c643f791bae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101132607
